Port IPC::SharedBufferReference to the new IPC serialization format

cdumez opened the pull request:
#### e3633fb4dca3f877787ad9a6518db345e11215a5
<pre>
Port IPC::SharedBufferReference to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=264957">https://bugs.webkit.org/show_bug.cgi?id=264957</a>

Reviewed by Alex Christensen.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/SharedBufferReference.cpp:
(IPC::SharedBufferReference::SharedBufferReference):
(IPC::SharedBufferReference::bufferHandle const):
(IPC::SharedBufferReference::encode const): Deleted.
(IPC::SharedBufferReference::decode): Deleted.
* Source/WebKit/Platform/IPC/SharedBufferReference.h:
(IPC::SharedBufferReference::buffer const):
* Source/WebKit/Platform/IPC/SharedBufferReference.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/270893@main">https://commits.webkit.org/270893@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10247fa5d81ab559482f703d5de0f9522be9a7bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26672 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5284 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27911 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24392 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7103 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2682 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24303 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26933 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4104 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22896 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3608 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3645 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29369 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24318 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24287 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29914 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3684 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1879 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27813 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5132 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6423 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4157 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4035 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Platform/IPC/SharedBufferReference.serialization.in:
```diff
@@ -0,0 +1,33 @@
+#if USE(UNIX_DOMAIN_SOCKETS)
+    RefPtr<WebCore::FragmentedSharedBuffer> buffer();
+#else
+    bool isNull();
```
This shouldn't be encoded separately.  This is somewhat included in whether bufferHandle() returns nullopt.  We used to have an attribute indicating that nullness was indicated by the first decoded bool, but we found it wasn't necessary with std::optional.

cdumez replied:
I'll look again but what you suggest if what I had in an earlier iteration and I had to move to a separate boolean to distinguish the null vs empty case since they're represented differently.

achristensen07 replied:
Having isNull=true and non-empty other pieces could cause unexpected behavior.  I think we may want to serialize a std::optional<SharedBufferReferenceData> rather than a bool AND other things

achristensen07 commented on Source/WebKit/Platform/IPC/SharedBufferReference.serialization.in:
```diff
@@ -0,0 +1,33 @@
+    RefPtr<WebCore::FragmentedSharedBuffer> buffer();
+#else
+    bool isNull();
+    size_t size();
```
I think we should make the size range check a validator instead of successfully decoding an object that we make empty.

cdumez replied:
How? We need to ::map() the handle before we can check the size of the mmap'd memory. What we decode is a handle.

achristensen07 replied:
Hmmm.  I think you may be right.

cdumez committed:
Port IPC::SharedBufferReference to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=264957

Reviewed by Alex Christensen.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/SharedBufferReference.cpp:
(IPC::SharedBufferReference::SharedBufferReference):
(IPC::SharedBufferReference::bufferHandle const):
(IPC::SharedBufferReference::encode const): Deleted.
(IPC::SharedBufferReference::decode): Deleted.
* Source/WebKit/Platform/IPC/SharedBufferReference.h:
(IPC::SharedBufferReference::buffer const):
* Source/WebKit/Platform/IPC/SharedBufferReference.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/270893@main
